src/web/app/pages-instructor/instructor-course-enroll-page/instructor-course-enroll-page.component.ts

```diff
@@ -94,59 +99,129 @@ export class InstructorCourseEnrollPageComponent implements OnInit {
+            this.allStudentChunks.push(currentStudentChunk);
+            currentStudentChunk = [];
+          }
+        });
+    if (currentStudentChunk.length > 0) {
+      this.allStudentChunks.push(currentStudentChunk);
+      currentStudentChunk = [];
+    }
+
+    const enrolledStudents: Student[] = [];
+    this.remainingStudentChunks = [...this.allStudentChunks];
+
+    const enrollRequests: Observable<Students> = concat(
+        ...this.allStudentChunks.map((studentChunk: StudentEnrollRequest[]) => {
+          const studentsEnrollRequest: StudentsEnrollRequest = {
+            studentEnrollRequests: studentChunk,
+          };
+          return this.studentService.enrollStudents(
+              this.courseId, studentsEnrollRequest,
+          );
+        }),
+    );
+
+    this.handleTimeoutForEnrollRequest(enrollRequests).subscribe({
+      next: (resp: Students) => {
+        enrolledStudents.push(...resp.students);
+        this.remainingStudentChunks.shift();
+      },
+      complete: () => {
+        this.showEnrollResults = true;
+        this.statusMessage.pop(); // removes any existing error status message
+        this.statusMessageService.showSuccessToast('Enrollment successful. Summary given below.');
+
+        const allStudentEntries: StudentEnrollRequest[] = [];
+        this.allStudentChunks.forEach((chunk: StudentEnrollRequest[]) => allStudentEntries.push(...chunk));
+
+        this.enrollResultPanelList =
+            this.populateEnrollResultPanelList(this.existingStudents, enrolledStudents, allStudentEntries);
+
+        this.studentService.getStudentsFromCourse({ courseId: this.courseId }).subscribe((resp: Students) => {
+          this.existingStudents = resp.students;
+          if (!this.isExistingStudentsPanelCollapsed) {
+            const existingStudentTable: Handsontable = this.hotRegisterer.getInstance(this.existingStudentsHOT);
+            this.loadExistingStudentsData(existingStudentTable, this.existingStudents);
+          }
+          this.isExistingStudentsPresent = true;
+        });
+      },
+      error: (resp: ErrorMessageOutput) => {
+        this.enrollErrorMessage = resp.error.message;
+      },
     });
   }
 
+  private handleTimeoutForEnrollRequest(prevRequests: Observable<Students>): Observable<Students> {
+
+    return prevRequests.pipe(finalize(() => this.isEnrolling = false))
+        .pipe(takeWhile(() => this.remainingStudentChunks.length > 0))
+        .pipe(catchError((resp: ErrorMessageOutput) => {
+          // First request in the list caused the timeout
+          const currentLongRequest: StudentEnrollRequest[] = this.remainingStudentChunks[0];
+
+          // If the length of the long request falls under max / 4, abort the whole stream
+          // In other words, we limit the number of retries for a request to 2
+          if (currentLongRequest.length === this.numberOfStudentsPerRequest / 4) {
```

Comment:
It feels unsafe to use exact comparison. Can you make it `<=` or so? This will be future proof as if someone change the `numberOfStudentsPerRequest` to some odd number like 199 or so, the logic still hold no matter what (at most 1 more retry will happen)